draft-ietf-tls-esni.md

```diff
@@ -802,31 +808,21 @@ client-specific retry configurations in order to identify the client in a
 subsequent ECH handshake.
 
 If none of the values provided in "retry_configs" contains a supported version,
-the client can regard ECH as securely disabled by the server. As below, it
-SHOULD then retry the handshake with a new transport connection and ECH
-disabled.
-
-If the field contains any other value, the client MUST abort the connection with
-an "illegal_parameter" alert.
-
-If the server negotiates an earlier version of TLS, or if it does not provide an
-"encrypted_client_hello" extension in EncryptedExtensions, the client proceeds
-with the handshake, authenticating for ECHConfig.contents.public_name as
-described in {{auth-public-name}}. If an earlier version was negotiated, the
-client MUST NOT enable the False Start optimization {{RFC7918}} for this
-handshake. If authentication or the handshake fails, the client MUST return a
-failure to the calling application. It MUST NOT treat this as a secure signal to
-disable ECH.
+the client can regard ECH as securely disabled by the server.
+
+If an earlier TLS version was negotiated, the client can regard ECH as
+securely disabled by the server, the client MUST NOT enable the False Start
+optimization {{RFC7918}} for this handshake, and the client MUST abort the
+connection with an "ech_required" alert.
```

Comment:
```suggestion
securely disabled by the server. If the client retries, the client MUST NOT
enable the False Start optimization {{RFC7918}} for this handshake, and the
client MUST abort the connection with an "ech_required" alert.
```

Reply:
> securely disabled by the server. If the client retries, the client MUST NOT

I had been under the impression that the prohibition on RFC7918 use was for the handshake that contains the ECH rejection (that is, don't send application data into a handshake that is going to be aborted once it completes).  It's not clear why this prohibition would be conditioned on the client retrying the connection, or now whether it is meant to apply only to a retry handshake?

Reply:
The existing text already predicated RFC7918 on whether or not the client retried. (You can't do RFC7918 if you don't retry in this instance, that is.) This suggestion is meant as an editorial improvement (but YMMV), not as a semantic one.

Reply:
So I think we're understanding this the same way - that the point is not to plow through with RFC7918 in a connection that is being aborted due to ECH rejection, but the client is free to do so on the retry connection.  My mileage is definitely still varying on the "if the client retries", as it's really outside of the context of 'this handshake', which is what that part is about.

Would you say this would be equivalent?

```suggestion
securely disabled by the server, the client MUST NOT enable the False Start
optimization {{RFC7918}} for this handshake (it may be enabled for the retry
connection), and the client MUST abort the connection with an "ech_required" alert.
```


I'm going to push a commit shortly that I think irons out the ordering issue you flagged.

Reply:
Yep! (And reading that pointed out the confusion I introduced, so, thanks for clarifying.)

Reply:
The new commit I just pushed I think incorporates everything we've talked about (and is actually a bit more compact in the end) except that parenthetical about RFC7918 on the retry connection was still an open question when I did the commit.  If you think the new commit is otherwise good and that parenthetical will clarify it further, I'm all for it.